### PR TITLE
Tianxinliu/20230501

### PR DESCRIPTION
**Pull request by tianxin-ms**
## ARM API Information (Control Plane)

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow.
</i>
Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [x] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json [hunk 1]
```diff
@@ -1480,36 +1615,278 @@
...
+        },
+        "x-ms-pageable": {
+          "nextLinkName": "nextLink"
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CognitiveServices/accounts/{accountName}/raiPolicies/{raiPolicyName}": {
+      "get": {
+        "tags": [
+          "RaiPolicies"
```
Comment (yangyuan) [inline specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json:1698]: Some tags are RaiPolicy, some are RaiPolicies.

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json [hunk 2]
```diff
@@ -3396,7 +3833,15 @@
...
         "version": {
           "type": "string",
-          "description": "Deployment model version."
+          "description": "Optional. Deployment model version. If version is not specified, a default version will be assigned. The default version is different for different models and might change when there is new version available for a model. Default version for a model could be found from list models API."
+        },
+        "source": {
+          "type": "string",
+          "description": "Optional. Deployment model source ARM resource ID."
+        },
+        "source": {
```
Comment (yangyuan) [inline specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json:3842]: Duplicate?

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json [hunk 3]
```diff
@@ -3692,6 +4146,83 @@
...
+      "properties": {
+        "name": {
+          "type": "string",
+          "description": "The name of the model SKU."
+        },
+        "usageName": {
+          "type": "string",
+          "description": "The usage name of the model SKU."
+        },
+        "unitsPerInstance": {
```
Comment (yangyuan) [inline specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json:4203]: This contact has changed. https://msazure.visualstudio.com/Cognitive%20Services/_git/Platform-ResourceProvider?path=%2Fsrc%2FCommon%2FResourceProvider.Contracts%2FDeployments%2FModelSku.cs&_a=contents&version=GBmaster

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json [hunk 4]
```diff
@@ -3692,6 +4146,83 @@
...
+          "format": "int32",
+          "description": "The max capacity allowed."
+        },
+        "rateLimits": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/CallRateLimit"
+          },
+          "x-ms-identifiers": [
+            "rules"
```
Comment (yangyuan) [inline specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json:4219]: I dont think rules can be an identifier as rules are array

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json [hunk 5]
```diff
@@ -3813,6 +4358,175 @@
...
+          "readOnly": true
+        },
+        "etag": {
+          "type": "string",
+          "readOnly": true,
+          "description": "Resource Etag."
+        },
+        "properties": {
+          "$ref": "#/definitions/RaiPolicyProperties",
+          "description": "Properties of Cognitive Services account deployment."
```
Comment (yangyuan) [inline specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json:4397]: not deployment

**Comment by sjanamma** [comment]
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CognitiveServices/accounts/{accountName}/raiPolicies/{raiPolicyName}/effectivePolicy": {

Get on this resource is not allowed. Is effectivePolicies a Resource Type?


Get would be -> /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CognitiveServices/accounts/{accountName}/raiPolicies/{raiPolicyName}/effectivePolicies/{eggectivePolicyName}"
List Get would be -> /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CognitiveServices/accounts/{accountName}/raiPolicies/{raiPolicyName}/effectivePolicies


---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json:1410 in 660c281. [](commit_id = 660c28170eab2c8edf5fe9a746d4eee5eaf9f8b9, deletion_comment = False)

**Comment by sjanamma** [comment]
>         }

List call must support pagination.
Please add
  "x-ms-pageable": {
          "nextLinkName": "nextLink"
        }

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json:1692 in 660c281. [](commit_id = 660c28170eab2c8edf5fe9a746d4eee5eaf9f8b9, deletion_comment = False)

**Comment by sjanamma** [comment]
>           "type": "boolean",

ARM recommends enums over booleans for future proof APIs.

Standard guidance is: replace boolean/switch properties with a more meaningful enum whenever possible.
A boolean will forever have two valid values (true or false). A string enum type is always preferred. Also, properties should always provide better values just than True and False. For example two switches "isTypeA" and "isTypeB" should be replaced with one enum "type": [A, B, DefaultType]. Enums are always a more flexible and future proof option because they allow additional values to be added in the future in a non-breaking way, e.g. [Enabled, Disabled, Suspended, Deallocated].
Note: do NOT define a 'boolean enum' with two values 'True and False'. This might be easier to 'extend' in terms of types, but semantically its cofusing, and no better than a boolean.


---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json:2652 in 660c281. [](commit_id = 660c28170eab2c8edf5fe9a746d4eee5eaf9f8b9, deletion_comment = False)

**Comment by sjanamma** [comment]
>           "type": "boolean",

we recommend enums over booleans

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json:4258 in 660c281. [](commit_id = 660c28170eab2c8edf5fe9a746d4eee5eaf9f8b9, deletion_comment = False)

File: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json [hunk 6]
```diff
@@ -3379,6 +3799,23 @@
...
+          "type": "string",
+          "description": "Deployment model version upgrade option.",
+          "enum": [
+            "OnceNewDefaultVersionAvailable",
+            "OnceCurrentVersionExpired",
+            "NoAutoUpgrade"
+          ],
+          "x-ms-enum": {
+            "name": "DeploymentModelVersionUpgradeOption",
+            "modelAsString": true
```
[the remark this answers is not in the document]
Comment (tianxin-ms) [reply specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json:3817]: Confirmed with Yuan, we should have "modelAsString"

**Review by xiaomin-ms — APPROVED** [review]

**Review by TonyDou — APPROVED** [review]

**Comment by tianxin-ms** [comment]
@sjanamma I already fixed following your comments, can you please help us review it again?

**Comment by tianxin-ms** [comment]
> > ```
> > "/subscriptions/{subscriptionId}/providers/Microsoft.CognitiveServices/locations/{location}/models": {
> > ```
> 
> Looks like you have not modelled this as standard ARM resource .
> 
> Is this a standard contract that you are implementing like usages ? Can you please share any documentation on this ?
> 
> Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json:888 in [e5fccc9](https://github.com/Azure/azure-rest-api-specs/commit/e5fccc97bc630e02014a3afcbeaf94c4f0f9e0b2). [](commit_id = [e5fccc9](https://github.com/Azure/azure-rest-api-specs/commit/e5fccc97bc630e02014a3afcbeaf94c4f0f9e0b2), deletion_comment = False)

@raosuhas , "models" here means AI models, not an ARM resource. This API means to get the list of AI models.

**Comment by xiaomin-ms** [comment]
> > ```
> > "/subscriptions/{subscriptionId}/providers/Microsoft.CognitiveServices/locations/{location}/models": {
> > ```
> 
> Looks like you have not modelled this as standard ARM resource .
> 
> Is this a standard contract that you are implementing like usages ? Can you please share any documentation on this ?
> 
> Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json:888 in [e5fccc9](https://github.com/Azure/azure-rest-api-specs/commit/e5fccc97bc630e02014a3afcbeaf94c4f0f9e0b2). [](commit_id = [e5fccc9](https://github.com/Azure/azure-rest-api-specs/commit/e5fccc97bc630e02014a3afcbeaf94c4f0f9e0b2), deletion_comment = False)

The models is read-only metadata for OpenAI models, it's very similar to "skus" API, so we don't include the "id", "type" as a regular ARM resource.

**Comment by v-jiaodi** [comment]
@tianxin-ms Please fix API Readiness Check error.

**Comment by v-jiaodi** [comment]
@msyyc @raych1 @qiaozha Could you help to take a look at the SDK Breaking Change error in this PR? Thanks

**Comment by qiaozha** [comment]
removed the JS breaking change label as there are no breaking changes.

**Review by v-jiaodi — APPROVED** [review]

**Comment by weidongxu-microsoft** [comment]
If you request SDK, comment in request when you think rollout be complete (so that user won't get SDK that not works).

**Pull request merged**